### PR TITLE
Add python-constraint to avoid conda-forge dependencies.

### PR DESCRIPTION
**Pull request by HackerFoo**
Any packages with a `condarc` specifying outside channels will require those channels to be available to install them.

Depending in conda-forge is undesirable because it can conflict with many of the default packages. `pip` cannot be used directly in a conda recipe, so PyPI packages must be converted to Conda packages.

This has caused problems with `vtr-optimized` ([PR](https://github.com/SymbiFlow/symbiflow-arch-defs/pull/1892)), which this PR aims to resolve, but there are other packages using `condarc` as well.

### CODE REVIEW

**Comment by PiotrZierhoffer** [comment]
Ha, we just discussed adding this, like minutes ago.

LGTM for me, but I'll ask @ajelinski to take a look as well

File: .travis.yml [hunk 1]
```diff
@@ -26,6 +26,11 @@ jobs:
     dist: xenial
     env:
     - PACKAGE=misc/magic
+  - stage: "No dependencies"
+    os: linux
+    dist: xenial
+    env:
+    - PACKAGE=misc/python-constraint
```
Comment (PiotrZierhoffer) [inline .travis.yml:33]: this should, probably, go to litex-conda-misc, but since it's a direct dependency of vtr-optimized, I don't think it's a big problem to have it here

File: .travis.yml [hunk 2]
```diff
@@ -183,7 +188,7 @@ jobs:
     dist: xenial
     env:
     - PACKAGE=pnr/vtr-gui
-  - stage: "No dependencies"
+  - stage: "Has first level dependencies"
```
Comment (litghost) [inline .travis.yml:191]: I don't think you meant to move every package into first level deps?

Comment (litghost) [reply .travis.yml:191]: Nevermind

**Review by litghost — APPROVED** [review]
LGTM

File: misc/python-constraint/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,51 @@
+{% set name = "python-constraint" %}
```
Comment (PiotrZierhoffer) [inline misc/python-constraint/meta.yaml:1]: is this file autogenerated with something like `conda skeleton pypi {packagename}`? I actually never used it before, but if this is the result, then it's pretty good

**Review by PiotrZierhoffer — APPROVED** [review]

**Comment by acomodi** [comment]
There should be another way to fix the python-constraint problem. Basically the condarc file in `vtr-optimized` should include also the `defaults` package.

All the dependencies for vtr-optimized will be taken from `defaults` apart from python-constraint which will be taken from conda-forge. Given that python-constraint does not have any other dependency itself, it should not generate any conflicts.

So basically the condarc file would look like the following:
```
channels:
  - defaults
  - conda-forge
```

in the current implementation of the vtr-optimized package, all the various dependencies are taken from conda-forge, hence the conflicts

File: misc/python-constraint/meta.yaml (hunk 3 above)
Comment (HackerFoo) [reply misc/python-constraint/meta.yaml:1]: Yeah, but I had to edit a bit, and change from .gz to .bz2.

**Comment by HackerFoo** [comment]
> There should be another way to fix the python-constraint problem. Basically the condarc file in `vtr-optimized` should include also the `defaults` package.
> 
> All the dependencies for vtr-optimized will be taken from `defaults` apart from python-constraint which will be taken from conda-forge. Given that python-constraint does not have any other dependency itself, it should not generate any conflicts.
> 
> So basically the condarc file would look like the following:
> 
> ```
> channels:
>   - defaults
>   - conda-forge
> ```
> 
> in the current implementation of the vtr-optimized package, all the various dependencies are taken from conda-forge, hence the conflicts

@acomodi  I tried this, based on [this tip](https://conda-forge.org/docs/user/tipsandtricks.html), except reversed to avoid conda-forge, but unfortunately, it didn't work. There were still dependencies from conda-forge in `info/about.json`.

**Comment by HackerFoo** [comment]
@ajelinski Any comments?

**Comment by ajelinski** [comment]
Are you sure you tried with channels in this exact order? The order matters and in the link provided `conda-forge` is the most important channel.

I've tested how the recipe is rendered after adding the `defaults` channel on top in `condarc` like @acomodi mentioned. It definitely solves the problem which is caused by requiring libraries that are not available on the `defaults` (ie. `main`) channel:
```
  run:
  - libstdcxx-ng >=9.3.0
  - libgcc-ng >=9.3.0
  - tbb >=2020.2
```
This is because conda-build gets all packages it needs from `conda-forge` and there is gcc/gxx 9.3.0. After making `defaults` the most important channel (the first one in `condarc`) it becomes:
```
  run:
  - libstdcxx-ng >=7.3.0
  - tbb >=2020.3
  - libgcc-ng >=7.3.0
```
because conda-build will use all packages it can from the `defaults` channel. Only those not available will be taken from the `conda-forge` channel.

Those requirements can be satisfied without `conda-forge` so it won't be needed to install `vtr-optimized` after such change.

**Comment by ajelinski** [comment]
Btw. `info/about.json` doesn't matter during dependency resolving. The `depends` key from `info/index.json` inside the package is what matters the most and there you can see this exact problem with requiring `libgcc/libstdcxx >=9.3.0`.

**Comment by ajelinski** [comment]
I have added a branch with the aforementioned `condarc` change and all run requirements of the built package can be satisfied without `conda-forge`: https://travis-ci.com/github/litex-hub/litex-conda-eda/jobs/461371641#L2925

The package built in that CI can be installed with:
```
conda install -c litex-hub/label/travis-fix-vtr-optimized-209189450 vtr-optimized
```

**Comment by HackerFoo** [comment]
@ajelinski Thanks. I'll try you package and close this if it works, unless there is still value in it.

**Comment by HackerFoo** [comment]
@ajelinski It seems to work, so made a PR with your branch: https://github.com/litex-hub/litex-conda-eda/pull/62

**Comment by HackerFoo** [comment]
Superseded by https://github.com/litex-hub/litex-conda-eda/pull/62
